### PR TITLE
#369 - Work Package Form Validation

### LINKED ISSUE #362: Projects - Work Package Edit Form
## Description

A new component needs to be created so that project managers can edit a work package. This component will be a form that will look similar to the work package details page, except that all the fields within within the work package will be editable text boxes rather than view-only details about the work package. 

## Background
The idea is that when project managers click on the "edit" button (which will be completed in another ticket), this form will be populated with the current information of the work package, so that project managers can directly edit and save the details. This ticket is to make the form that the project managers will see when they click the "edit" button. Do not worry about populating the text boxes with the current values, as that will be completed in another ticket. The focus of this ticket is to only create the front-end visual.

## Proposed Solution
\*\***For all sections listed below, the headers for each section (Work Package Details, Dependencies, etc.) should not be editable**\*\*

### Work Package Details 
- [x] The labels for each field will remain exactly the same and in the same location. Only the details portion for each field will change and will become input text boxes. 
- [x] The "Start Date" field should become a date picker.
- [x] The "End Date" field should be read only, since it is calculated by the start date and duration.
- [x] The percent symbol for the "Progress" and "Expected Progress" fields should be a read-only field. The user will only be changing the numbers. See mocks for clarification.
- [x] The "weeks" unit for the "Duration" field should be a read-only field. The user will only be changing the numbers. See mocks for clarification.
- [x] The "status" field on the top right corner of this section should be a dropdown list with the following three selections: "Inactive", "Active", "Complete".

### Dependencies
- [x] Each existing dependency will become a text box with a delete button on the upper left hand corner of each text box. 
- [x] At the end of the list, there should be a blank text box with an add button on the right hand side of the text box. This text box and button should always be present, whether or not there are any list items present.

### Expected Activities
- [x] Each existing expected activity will become a text box with a delete button for each line item located on the right hand side of the screen.
- [x] At the end of the list, there should be a blank text box with an add button on the right hand side of the text box. This text box and button should always be present, whether or not there are any list items present.

### Deliverables
- [x] Each existing deliverable will become a text box with a delete button for each line item located on the right hand side of the screen.
- [x] At the end of the list, there should be a blank text box with an add button on the right hand side of the text box. This text box and button should always be present, whether or not there are any list items present.

### Changes
- [x] View only section. The section should not be editable. It should still be a part of the form, but should remain a view-only section.

### General 
- [x] There should be a sticky footer with two buttons: "Cancel" and "Save". The footer will be slightly transparent.

## Mocks
https://www.figma.com/file/qmXuySSwfiD942HwmSSEMe/NERPM-Desktop?node-id=392%3A225

### LINKED ISSUE #369: Projects - Work Package Form Validation
## Description
When implementing change requests, project managers will edit a form populated with the information from a work package. We need to implement front-end validation to ensure that the user inputs and submits proper data. Having front-end validation helps ensure that the information is correctly formatted.

## Background
The application does not currently support any front-end validation. Including this in the work package edit form will help ensure that the submitted data is consistent with the formatting in our database. Validation will be similar to the ones in #370.

### Work Package Details
- [x] The "Duration" and "Progress" fields should contain only numbers
- [x] WBS # and "Expected Progress" fields: none because it is read-only

### Dependencies
- [x] All elements pass the `validateWBS()` function provided by the `utils` package

### Expected Activities
- [x] Ensure that the section is not empty

### Deliverables
- [x] Ensure that the section is not empty

### Changes
- [x] No form validation needed because this section should be read-only

### PR DESCRIPTION
**Pull request by z-roth**
This ticket is not complete - opening for review.

**Completed**
-Duration and progress fields can already only contain numbers as they are number inputs. I set the minimum value to 0 as they will not be negative.
-WBS# and Expected Progress are read-only, and have no validation or editability.
-None of the bullet list items (Expected Activities, Deliverables) can have an empty textbox (unless it is the add new bullet textbox)

**Uncompleted**
- [x]  Validation for Dependencies. Stuck on trying to figure out an implementation for it - will use try and catch with ValidateWBS, but working on creation of an on-screen validation error message.
- [x]  Create function to reformat startDate to "YYYY-MM-DD" so it can be used as a default value for the date.

**Notes/Questions**
- Should we remove the capability of the CR ID being picked by a user? Adds an extra step to validation - having to make the async call to check if it is taken. Instead we can make it randomly generated when submitted.
- Dependencies can only be added once at a time with current implementation? Does + button add a new dependency input?

Resolves #369

### CODE REVIEW

**Comment by jpwaves** [comment]
I've generally handled CR ID validation by just making sure the number inputted could potentially be a CR ID and letting the backend handle verifying that the number given is a valid CR ID, which is not the best way to do validation but it seemed the fastest way to get it done. As far as dependencies go, adding a new dependency should create a new input box for it and dependencies I believe are added one at a time. If you're using the `EditableTextInputList` component for getting the list of dependencies in the form, all the dependencies would be stored in a state variable with the dependencies represented as a list of strings.

File: src/components/shared/editable-text-input-list/editable-text-input-list.tsx [hunk 1]
```diff
@@ -3,37 +3,65 @@
...
+    </Button>
+  );
+
+  if (!readOnly) {
+    listPrepared = [...listPrepared, addButton];
+  }
+  let builtList = <ul>{listPrepared}</ul>;
+  if (ordered) {
+    builtList = <ol>{listPrepared}</ol>;
+  }
```
Comment (jamescd18) [inline src/components/shared/editable-text-input-list/editable-text-input-list.tsx:61]: I don't think we need nor want to have HTML list tags in this stuff. If it isn't in there for the project edit form then let's not add it for this.

It looks like unnecessary indentation to me
Screen Shot 2022-03-14 at 11 06 22 PM

**Review by jamescd18 — CHANGES_REQUESTED** [review]
This is amazing work! There's a few different things I think need to change before it's ready to go, but it looks to be most of the way there. Also, be sure to use the new PR template before pinging for a second review. Awesome job!

**Comment by jamescd18** [comment]
## Changes

_Explanation of changes goes here_

## Notes

_Any other notes go here_

## Test Cases

- Case A
- Edge case
- ...

## Screenshots (if applicable)

_Any relevant screenshots go here_

## To Do

_Any remaining things that need to get done_

- [ ] item 1
- [ ] ...

## Checklist

It can be helpful to check the `Checks` and `Files changed` tabs.
Please review the [contributor guide](https://github.com/Northeastern-Electric-Racing/PM-Dashboard-v2/blob/main/docs/ContributorGuide.md) and reach out to your squad if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [ ] All commits are tagged with the ticket number
- [ ] No linting errors
- [ ] No newline at end of file warnings
- [ ] All code follows repository-configured prettier formatting
- [ ] No merge conflicts
- [ ] All checks passing
- [ ] Screenshots of UI changes (if applicable)
- [ ] Remove any not-applicable sections
- [ ] Assign the PR to yourself
- [ ] PR is linked to the ticket
- [ ] No `package-lock.json` changes (unless dependencies have changed)
- [ ] Request reviewers & ping on Slack

Closes # (issue #)

File: src/components/projects/wbs-details/work-package-container/work-package-container.tsx [hunk 2]
```diff
@@ -30,24 +35,105 @@ export interface EditModeProps {
 const WorkPackageContainer: React.FC<WorkPackageContainerProps> = ({ wbsNum }) => {
   const { isLoading, isError, data, error } = useSingleWorkPackage(wbsNum);
   const [editMode, setEditMode] = useState(false);
+  const [form, setForm] = useState({});
+  const [expectedActivities, setExpectedActivities] = useState(['']);
+  const [deliverables, setDeliverables] = useState(['']);
+
+  // On render, data is undefined. This causes issues if trying to set the state to fields from it. Bug?
```
[the remark this answers is not in the document]
Comment (z-roth) [reply src/components/projects/wbs-details/work-package-container/work-package-container.tsx:42]: Should I do anything else about this?

File: src/components/projects/wbs-details/work-package-container/work-package-details/work-package-details.test.tsx [hunk 3]
```diff
@@ -60,8 +57,8 @@ describe('Rendering Work Package Details Component', () => {
     expect(
       screen.getByText(`${endDatePipe(wp.startDate, wp.duration)}`, { exact: false })
     ).toBeInTheDocument();
-    const progresses = screen.getAllByText(`${percentPipe(wp.progress)}`); // progress and expectedProgress should be equal and return 2 results
-    expect(progresses.length).toBe(2);
+    const progresses = screen.getAllByText(`${percentPipe(wp.progress)}`); // expected progress has been removed, so only 1 progress at the moment.
```
[the remark this answers is not in the document]
Comment (z-roth) [reply src/components/projects/wbs-details/work-package-container/work-package-details/work-package-details.test.tsx:60]: Honestly, not really sure. I saw that this test was counting on Progress and Expected Progress being rendered, but it looks like ExpectedProgress doesn't exist anymore with one of the more recent pulls. This was causing tests to fail.

File: src/components/shared/editable-text-input-list/editable-text-input-list.tsx (hunk 1 above)
Comment (jpwaves) [reply src/components/shared/editable-text-input-list/editable-text-input-list.tsx:61]: I think it's fine if you want to turn this into a list HTML element, but I think a better way to go about it would be to replace the `div` element in the `map` with `li` and then wrap the output of the `map` in an `ul` so that it's something like `<ul>{mappedItems}</ul>`. Only thing with this is that you need to make sure to remove the default stylings from the list HTML elements (mainly `ul`, not sure if `li` also has default styling that needs to be reset)

File: src/components/shared/editable-text-input-list/editable-text-input-list.tsx [hunk 4]
```diff
@@ -3,37 +3,65 @@
...
+          />
+          <Button type="button" variant="danger" onClick={() => remove(index)}>
+            X
+          </Button>
+        </InputGroup>
+      </div>
+    ) : (
+      <li key={index}>{item}</li>
+    )
+  );
```
Comment (jpwaves) [inline src/components/shared/editable-text-input-list/editable-text-input-list.tsx:47]: `Form.Control` has a prop to make it readonly, so there's no need for the ternary operator. Also, if we're making the list readOnly, we'd probably want a ternary operator on the remove button instead (if `readOnly`, don't render anything aka `<></>`; otherwise, render the remove button)

File: src/components/shared/editable-text-input-list/editable-text-input-list.tsx [hunk 5]
```diff
@@ -3,37 +3,65 @@
...
   add: (val: any) => any;
   remove: (idx: number) => any;
   update: (idx: number, val: any) => any;
 }
 
 const EditableTextInputList: React.FC<EditableTextInputListProps> = ({
+  title,
   items,
+  readOnly,
+  ordered,
```
Comment (jpwaves) [inline src/components/shared/editable-text-input-list/editable-text-input-list.tsx:23]: I'm not sure how `ordered` would work in this type of list since the list items are text inputs and those can be changing at any time, so depending on what is in each text input at a given time, the order may be different. How were you envisioning `ordered` to work?

File: src/components/shared/bullet-list/bullet-list.tsx [hunk 6]
```diff
@@ -6,30 +6,71 @@
...
+          placeholder={bullet.props.children}
+          key={idx}
+          onChange={(e) => handleChange(`${fieldName}${idx}`, e.target.value)}
+        />
+        <Button variant="danger" key={idx} onClick={handleDelete}>
+          X
+        </Button>
       </InputGroup>
     ) : (
       <li key={idx}>{bullet}</li>
```
Comment (jpwaves) [inline src/components/shared/bullet-list/bullet-list.tsx:76]: Same comment and recommended edits as with the `EditableTextInputList`, `readOnly ` is a prop on `Form.Control` that you can use and we can probably take out the remove button rendering when it is read only.

File: src/components/projects/wbs-details/work-package-container/work-package-details/work-package-details.tsx [hunk 7]
```diff
@@ -9,7 +9,7 @@ import PageBlock from '../../../../shared/page-block/page-block';
 import { Col, Container, Row, Form } from 'react-bootstrap';
```
Comment (jpwaves) [inline src/components/projects/wbs-details/work-package-container/work-package-details/work-package-details.tsx:9]: idk why I can't select line 7, but Github is showing a warning that `percentPipe` isn't ever used.

File: src/components/projects/wbs-details/work-package-container/work-package-container.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (jpwaves) [reply src/components/projects/wbs-details/work-package-container/work-package-container.tsx:42]: I think how I resolved this is to make a wrapper component that gets the data separately and then passes that data to the view component (I presume that is this). That way, the view component will be guaranteed to always have data that is defined such that you show the loading indicator while you wait for the data, and then once the data is retrieved, you render the form with the data at the same time.

File: src/components/projects/wbs-details/work-package-container/dependencies-list/dependencies-list.tsx [hunk 8]
```diff
@@ -7,22 +7,57 @@ import HorizontalList from '../../../../shared/horizontal-list/horizontal-list';
...
+  }
+
+  function handleAdd() {
+    try {
+      setDependenciesState([...dependenciesState, validateWBS(unvalidatedDependency)]);
+    } catch (error: any) {
+      alert(error.message);
+    }
+    setUnvalidatedDependency('');
+  }
```
Comment (jpwaves) [inline src/components/projects/wbs-details/work-package-container/dependencies-list/dependencies-list.tsx:56]: @jamescd18 correct me if I'm wrong, but I think we generally try to use ES6 arrow function expressions over function declarations

File: src/components/shared/editable-text-input-list/editable-text-input-list.tsx [hunk 9]
```diff
@@ -3,37 +3,65 @@
...
-                update(index, e.target.value);
-              }} />
-              <Button type='button' variant='danger' onClick={() => remove(index)}>X</Button>
-            </InputGroup>
-          </div>
-        ))
-      }
-      <Button type='button' variant='success' onClick={() => add('')}>+ Add New Bullet</Button>
-    </>
+    <PageBlock title={title} headerRight={<></>} body={<Form.Group>{builtList}</Form.Group>} />
```
Comment (jpwaves) [inline src/components/shared/editable-text-input-list/editable-text-input-list.tsx:64]: I don't think it's a good idea to wrap the content of the `EditableTextInputList` in a `PageBlock`, reason being that it would make the styling very weird if we wanted to have multiple `EditableTextInputList`s in a single `PageBlock` or some other component. It does mean that there is more work to do in other higher-level components, but that also means this component can be easily swapped out in the future if we decide to make a new implementation of `EditableTextInputList` (which I hope we do because I made this very quickly so there are quite a few issues with it) or decide we don't want to use it anymore

**Review by jpwaves — CHANGES_REQUESTED** [review]
Just my initial thoughts on looking at the code, don't have the time tonight to manually test but I'll do that tomorrow

File: src/components/shared/editable-text-input-list/editable-text-input-list.tsx (hunk 5 above)
Comment (jamescd18) [reply src/components/shared/editable-text-input-list/editable-text-input-list.tsx:23]: I think I mentioned the whole list tag stuff should be removed

Comment (jpwaves) [reply src/components/shared/editable-text-input-list/editable-text-input-list.tsx:23]: Yeah, I saw that - semantically speaking, it would make sense to make this use a list tag as opposed to `div`s. Only caveat would be to remove all the default stylings on `li`, `ul`, and `ol` tags and then apply the styling that is currently on the `div`

File: src/components/projects/wbs-details/work-package-container/work-package-container.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (z-roth) [reply src/components/projects/wbs-details/work-package-container/work-package-container.tsx:42]: Is it alright if I keep it like this? It works, but I'm also a bit confused about how I could implement such a wrapper. Should I create a component up or down the tree?

File: src/components/shared/editable-text-input-list/editable-text-input-list.tsx (hunk 4 above)
Comment (z-roth) [reply src/components/shared/editable-text-input-list/editable-text-input-list.tsx]: True, but the readOnly style looks like a greyed-out textbox. Do we want this kind of styling? From a UI/UX standpoint I feel like it would be a little confusing.

File: src/components/shared/editable-text-input-list/editable-text-input-list.tsx (hunk 5 above)
Comment (z-roth) [reply src/components/shared/editable-text-input-list/editable-text-input-list.tsx:23]: When I implemented this, I was looking at the previous implementation (the BulletList component) to see how I could port it over to the complete EditableTextInputList. Should I remove this or keep it in?

File: src/components/shared/editable-text-input-list/editable-text-input-list.tsx (hunk 9 above)
Comment (z-roth) [reply src/components/shared/editable-text-input-list/editable-text-input-list.tsx:66]: Thoughts on adding a PageBlock prop? So it can be rendered in it's own pageblock or without it

File: src/components/shared/editable-text-input-list/editable-text-input-list.tsx (hunk 4 above)
Comment (jpwaves) [reply src/components/shared/editable-text-input-list/editable-text-input-list.tsx]: Not really sure tbh, I'd ask the PMs about that in the product-request or product channel. Although, this is just MVP so I've been operating under the impression that UI doesn't need to be perfect and so I've let things like this just slide.

File: src/components/shared/editable-text-input-list/editable-text-input-list.tsx (hunk 5 above)
Comment (jpwaves) [reply src/components/shared/editable-text-input-list/editable-text-input-list.tsx:23]: If it's easier for you to keep it, I think it's fine to keep it. We can always make a ticket (and I hope we do) to refine EditableTextInputList in the future

File: src/components/shared/editable-text-input-list/editable-text-input-list.tsx (hunk 9 above)
Comment (jpwaves) [reply src/components/shared/editable-text-input-list/editable-text-input-list.tsx:66]: If you add a PageBlock prop, that would be tightly coupling this component to the PageBlock component, which is what I'm trying to avoid. Even if we end up continuing to use PageBlock's for our UI, I think it's still bad design to have components reliant on each other in this manner, especially with how versatile the use cases for an EditableTextInputList can be.

File: src/components/projects/wbs-details/work-package-container/work-package-container.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (jpwaves) [reply src/components/projects/wbs-details/work-package-container/work-package-container.tsx:42]: If you went with a wrapper component, you'd make a component up the tree. It would do all the data processing and contain all the business logic like handler functions and things of that sort, and then pass that all down to another component (which I believe is what this is) that acts as a view. There are quite a few examples of this pattern throughout the codebase -- a simple example of this would be the review change request form if you want to check that out.

File: src/components/shared/editable-text-input-list/editable-text-input-list.tsx (hunk 4 above)
Comment (z-roth) [reply src/components/shared/editable-text-input-list/editable-text-input-list.tsx]: Sent a message in #product-request.

Comment (anthonybernardi) [reply src/components/shared/editable-text-input-list/editable-text-input-list.tsx]: personally I prefer the <li> and it doesn't really add any complexity i think

**Comment by z-roth** [comment]
## Changes

- Implemented date formatter for default date value for EditableDetail
- Implemented add, delete, and validate functionalities for Dependency component.
- Added min and max fields to editable details of number type
- Implemented EditableTextInputList for Work Package Edit Form, added readOnly functionality
- Converted handleAdd and handleDelete functions to arrow functions
- Created view component for work package container

## Notes

## Test Cases

- No new testing due to no new major UI components.

## Screenshots (if applicable)

_Any relevant screenshots go here_

## To Do

- [ ] Fix failing tests

## Checklist

It can be helpful to check the `Checks` and `Files changed` tabs.
Please review the [contributor guide](https://github.com/Northeastern-Electric-Racing/PM-Dashboard-v2/blob/main/docs/ContributorGuide.md) and reach out to your squad if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [ ] All commits are tagged with the ticket number
- [ ] No linting errors
- [ ] No newline at end of file warnings
- [ ] All code follows repository-configured prettier formatting
- [ ] No merge conflicts
- [ ] All checks passing
- [ ] Screenshots of UI changes (if applicable)
- [ ] Remove any not-applicable sections
- [ ] Assign the PR to yourself
- [ ] PR is linked to the ticket
- [ ] No `package-lock.json` changes (unless dependencies have changed)
- [ ] Request reviewers & ping on Slack

Closes # (issue #)

**Review by jamescd18 — CHANGES_REQUESTED** [review]
Not sure exactly which test but it looks like the test is failing because the WBS # is different. one is 1.12.0 and the other is 1.1.1. So that's gotta be fixed one way or the other to avoid that situation.

Screen Shot 2022-03-24 at 9 30 29 PM

**Comment by z-roth** [comment]
@jpwaves @jamescd18 could one of you guys check out these merge conflicts? I'm a bit confused on how to fix this. @anthonybernardi and I took a look into it today at our meeting, but nothing worked.

File: src/components/projects/wbs-details/work-package-container/work-package-container.module.css [hunk 10]
```diff
@@ -1,5 +1,4 @@
 /*
  * This file is part of NER's PM Dashboard and licensed under GNU AGPLv3.
  * See the LICENSE file in the repository root folder for details.
- */
-
+ */
```
Comment (acd124) [inline src/components/projects/wbs-details/work-package-container/work-package-container.module.css:4]: This is also apparently a merge conflict, I would recommend letting prettier run on the file, so hopefully it would end up not changed

**Comment by acd124** [comment]
`src/components/projects/wbs-details/work-package-container/dependencies-list/dependency/dependency.test.tsx` this file, and the other file in the same folder it said something about the file having changes in main but being deleted in this branch, I'm not sure how best to fix it other than merging main and making the delete again, not sure what actually happened for those two files though

**Review by jamescd18 — CHANGES_REQUESTED** [review]
Can you please change the project lead and project manager fields so that they're instead dropdowns populated with the list of users? The logic should be available to understand and copy from the project edit form.

**Review by jamescd18 — APPROVED** [review]
LGTM! Soooo exciting!

**Comment by z-roth** [comment]
Woot!!! 🥳🍾
